Grafana RUM MVP

elipe17 opened the pull request:
## Summary of Changes
Frontend:
- Integrated FaroSDK
- Added custom hooks to further implement RUM into TDP
- Instrumented: user info, page changes, session tracking, request tracing/spans
- Updated nginx config to support faro

Backend:
- Instrumented open telemetry libs to track users requests and flows throughout the whole app and many services
- Added new middleware for telemetry
- Added ability to enable/disable telemetry and tracing
- Added Grafana Tempo to track trace/span storage for architecture visualization
- Update alloy configs to stage incoming faro metrics to store in loki


Notes:
- There are some oddities with faro tracking route change events. all urls are logged per event but the actual `route_change` event is not firing as frequently as it should.

Pull request closes #4780

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Change your route a few times, submit some files, etc...
3. Go Grafana and check out the OpenTelemetry dashboard

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] A frontend observability tool is selected and agreed upon
- [ ] MVP dashboard displays basic session metrics (session count, top pages, navigation flow)
- [ ] Data is captured in a way that supports user analysis while obscuring sensitive information
- [ ] Setup is reviewed for OCIO/compliance alignment (see [Issue #2006](https://github.com/raft-tech/TANF-app/issues/2006))
- [ ] README and documentation are updated as needed
- [ ] Testing Checklist has been run and all tests pass
- [ ] [Release digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) has been updated with testing steps & relevant media

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
Merge remote-tracking branch 'origin/3258-alloy-conversion' into 4780-grafana-rum-mvp

elipe17 committed:
- Initial commit

elipe17 committed:
- remove files

elipe17 committed:
- mvp tempo/alloy

elipe17 committed:
- Working impl of RUM

elipe17 committed:
- Add instrumentation for all available services
- Update user node tracking

elipe17 committed:
- linting

elipe17 committed:
- remove unnecessary config

elipe17 committed:
- Add dashboard

elipe17 committed:
- Working implementation of faroSDK with minimal grafana dashboard

elipe17 committed:
- Configure tempo to store data in S3

elipe17 committed:
- Add tempo datasource

elipe17 committed:
- remove topology

elipe17 committed:
- Small dashboard update

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 4780-grafana-rum-mvp

elipe17 committed:
- Remove redux middle ware
- Move service into a hook

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 4780-grafana-rum-mvp

elipe17 committed:
- Update base image for frontend

elipe17 committed:
- Update package lock

elipe17 committed:
- Re-update package-lock

elipe17 committed:
- Pushing as test

elipe17 committed:
- Re-updated packages after installing new node version

elipe17 committed:
Merge branch 'develop' into 4780-grafana-rum-mvp

elipe17 committed:
- resolve conflicts

elipe17 committed:
- Fix package-lock file
- Add ignore pattern for faro
- Update App tests to support redux selectors and routing
- Add mocks for faro resources

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 4780-grafana-rum-mvp

elipe17 committed:
- Remove file
- Ignore faro hook from coverage

elipe17 committed:
- remove unneeded checks

elipe17 committed:
- remove outdated readme

elipe17 committed:
- Add missing vars

elipe17 committed:
- sort by time

elipe17 committed:
- only show 95 percentile

elipe17 committed:
- show all values in legend

elipe17 committed:
- quick update

elipe17 committed:
- Add traces

elipe17 committed:
- revert dockerfile

elipe17 committed:
- remove unnecessary ports

lhuxraft commented:
@elipe17 I went to test this using the tdp-dev credentials and I don't see the OpenTelemetry dashboard -- are there other credentials I should be using?

elipe17 commented:
> @elipe17 I went to test this using the tdp-dev credentials and I don't see the OpenTelemetry dashboard -- are there other credentials I should be using?

This is local only. To deploy this requires deploying another new app called Tempo with our PLG stack. Happy to help you set it up locally for testing :)

lhuxraft commented:
Hi @elipe17! Thanks again for helping me get this set up on local :) 

This is a super great start and it's awesome to see all of this data in one place. Here are some of my suggestions for visualizations to help us get what we want out of this data -- we don't have to do all of this, but hopefully this gets your wheels turning on what we'd be looking for:
- Page loads / errors over time (visualization inspo [here](https://grafana.com/products/cloud/frontend-observability-for-real-user-monitoring/))
  - The default view should be at a high level, but the ability to drill down into specific components / data will be really important. Example drilldown capablities I'd like to see include time period, types of errors, types of users / specific user IDs, page URL, browser, etc. Additionally, I didn't see this in the original data but it may also be helpful to map user IDs to STTs / regions for additional filtering capabilities as well.
- KPIs / summary stats like time to first byte, first contentful paint, largest contentful paint, first input delay, # of page loads, etc
- Performance over time by page
- Most common user navigation paths and/or which pages receive the most traffic
- Geographical distribution of users
- Amount of time being spent on each page (with filters of course)
- Session counts over time
- Potentially: memory/CPU usage over time, logins over time, server requests over time, etc

Hope this is helpful and let me know if you have any questions! Happy to workshop through any of these if needed. Also, I'm happy to track these additions as part of the finalization ticket or here. Whatever your preference is :)

elipe17 committed:
- Added an overview dashboard for some of the requested visualizations

elipe17 commented:
> Hi @elipe17! Thanks again for helping me get this set up on local :)
> 
> This is a super great start and it's awesome to see all of this data in one place. Here are some of my suggestions for visualizations to help us get what we want out of this data -- we don't have to do all of this, but hopefully this gets your wheels turning on what we'd be looking for:
> 
> * Page loads / errors over time (visualization inspo [here](https://grafana.com/products/cloud/frontend-observability-for-real-user-monitoring/))
>   
>   * The default view should be at a high level, but the ability to drill down into specific components / data will be really important. Example drilldown capablities I'd like to see include time period, types of errors, types of users / specific user IDs, page URL, browser, etc. Additionally, I didn't see this in the original data but it may also be helpful to map user IDs to STTs / regions for additional filtering capabilities as well.
> * KPIs / summary stats like time to first byte, first contentful paint, largest contentful paint, first input delay, # of page loads, etc
> * Performance over time by page
> * Most common user navigation paths and/or which pages receive the most traffic
> * Geographical distribution of users
> * Amount of time being spent on each page (with filters of course)
> * Session counts over time
> * Potentially: memory/CPU usage over time, logins over time, server requests over time, etc
> 
> Hope this is helpful and let me know if you have any questions! Happy to workshop through any of these if needed. Also, I'm happy to track these additions as part of the finalization ticket or here. Whatever your preference is :)

@lhuxraft, I added a new `Telemetry Overview` dashboard that provides errors over time and by page, all requested KPIs, performance metrics over time, browsers incurring the most errors, most popular browsers, and the most common Faro events. This is an overview dashboard so it is system wide. I plan to try and get some of the user info in the next PR. I also updated our Grafana version. This enables the drilldown capabilities and adds a few new visualizations.

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 4780-grafana-rum-mvp

elipe17 committed:
- Updated grafana version to get drilldown capabilities and new visualizations

elipe17 committed:
Merge branch 'develop' into 4780-grafana-rum-mvp

jtimpe commented on tdrs-frontend/src/axios-instance.js:
```diff
@@ -1,9 +1,33 @@
+const axiosInstance = axios.create()
+
+// Add an interceptor to include trace context in outgoing requests
+axiosInstance.interceptors.request.use((config) => {
```
curious if this would miss any requests due to us not using the `axiosInstance` everywhere we make requests. similar to #3510

elipe17 replied:
Thats a good catch! Im sure it does miss requests, ill have to see if I can set this globally on axios and not just the custom instance

jtimpe commented on tdrs-frontend/src/components/Routes/Routes.js:
```diff
@@ -25,7 +26,7 @@ const AppRoutes = () => {
 
   return (
-    <Routes>
+    <FaroRoutes>
```
could be me missing something, but i don't see all route changes captured in the dashboard

elipe17 replied:
This is an issue. It happens because the frontend is mostly using link based page navigation as opposed to react router. The FaroRoutes only tracks react router based navigation. Thats why I added the hooks so we can manage it ourselves. I just didn't actually invoke the hooks anywhere because I already had so much data to manipulate.

jtimpe commented on tdrs-frontend/src/hooks/useRUM.js:
```diff
@@ -0,0 +1,121 @@
+      timestamp: Date.now(),
+    }
+
+    faro.api.pushEvent('page_view', eventData)
```
i don't believe i see any `page_view`s get generated either

elipe17 replied:
Correct, none of these functions from the hook are used except the `setUserInfo` function

jtimpe commented on tdrs-frontend/src/index.js:
```diff
@@ -57,7 +113,9 @@ const root = createRoot(container)
   <Provider store={store}>
     <Router store={store} history={history}>
-      <App />
+      <FaroErrorBoundary>
```
i noticed this exception get generated
![image](https://github.com/user-attachments/assets/1b11e868-ea8b-49ea-a69c-19a9a8fc86d0)

this was from when a file was added with the wrong submission year/quarter - the exception gets caught in our code but looks like it still generates a FARO event. also i'm afraid the stack trace/context provided won't be super useful when our codebase is deployed (since everything is minified) - i believe sentry addresses this somehow to allow frontend stack traces. maybe there's a FARO equivalent for the technique they use.

elipe17 replied:
That's an interesting error. As of now, Im not using the stacktrace field in the dashboards. Just counting the exceptions and what URL they occurred on. I will do some digging on how faro trys to manage stacktraces for production builds.

jtimpe reviewed (approved):
really good start! i noticed a couple small things that i think we will continue to iron out as we move forward. great work

mattcoleanderson reviewed (approved):
Great work! I didn't see anything worth commenting on.

elipe17 committed:
- point back to ghcr

elipe17 committed:
Merge branch 'develop' into 4780-grafana-rum-mvp

elipe17 committed:
- added interceptor for default axios insstance

elipe17 committed:
Merge branch '4780-grafana-rum-mvp' of https://github.com/raft-tech/TANF-app into 4780-grafana-rum-mvp

elipe17 committed:
Merge branch 'develop' into 4780-grafana-rum-mvp

elipe17 committed:
- Add tests to bolster coverage

elipe17 committed:
Merge branch 'develop' into 4780-grafana-rum-mvp
